types/wlr_output.c

```diff
@@ -706,6 +789,9 @@ static bool output_basic_test(struct wlr_output *output) {
 }
 
 bool wlr_output_test(struct wlr_output *output) {
+	if (!output_ensure_buffer(output)) {
```

Comment:
Do we want to attach buffers to the output before doing the basic output test? And do we even want wlr_output_test to attach buffers? Seems a bit intrusive.

Reply:
That's a good question. The reason a buffer is attached is that some backends might only be able to test a modeset if they have a buffer. This is the case of the DRM backend for instance, previously an empty black buffer was attached on modeset if none was provided by the compositor.

For the X11 and Wayland backends performing a modeset without a buffer doesn't really make sense either.

Reply:
Oh yeah, it's also needed to test the modeset.

However, does it make sense to perform `output_basic_test` *before* ensuring a buffer? This would let us bail out early from an obviously bad configuration without wasting time allocating and clearing buffers (possibly twice if we fall back to modifierless). I don't think `output_basic_test` relies on the output from `output_ensure_buffer`—the buffer tests have matching conditions in `output_ensure_buffer`, and the direct scan-out resolution test shouldn't be relevant if we had to allocate the buffer.

It should also be noted that we now issue up to 3 tests towards the output impl from `wlr_output_test`, as `output_ensure_buffer` performs 1 or 2 tests of its own. This shouldn't be a problem if the test is sufficiently fast.

Reply:
> However, does it make sense to perform output_basic_test before ensuring a buffer?

I've been wondering about this as well. Updated to allocate the buffer only if the basic test succeeds, we can always change this in the future if necessary.

Note, once the first test has allocated a swapchain+buffer, these are re-used in subsequent tests and commits, so the overhead shouldn't be too bad.